src/enclave/enclave_signal.c

```diff
@@ -143,6 +143,32 @@ static uint64_t sgxlkl_enclave_signal_handler(
             exception_record->address,
             opcode);
 
+        /**
+         * If LKL has not yet been initialised, we cannot handle the
+         * exception and fail instead.
+         */
+        if (!lkl_is_running())
+        {
+#ifdef DEBUG
+            sgxlkl_error("Unhandled exception. Printing stack trace before the signal handler was invoked:\n");
+            sgxlkl_print_backtrace((void*)oe_ctx->rbp);
```

Comment:
If I understand correctly this trace is from the ethread. Maybe make that clear in the output since you also mention "lthreads" in the other traces.

Reply:
It is the trace for the lthread, as all of our stacks are associated with lthreads. The difference here is that I use the frame pointer that was used when the exception happened, as opposed to the frame pointer used by the signal handler.

Reply:
OK, and dumping both is necessary because exception handling involves roundtrips that lose the original stacktrace? I'm trying to figure out what the difference to a normal environment is.

Reply:
I think that OE uses a separate stack for its exception handler inside the enclave. When I use the compiler generated frame pointer, I didn't see the lthread frames, which is why I retrieve it from the saved exception context. In contrast, `sgxlkl_print_backtrace()` iterates over the lthread scheduler state and dumps out the saved stacks for all lthreads.

I can add comments to explain this better.

Reply:
libUnwind can't unwind out of exception frames because exception frames don't contain DWARF unwind tables that tell them how to use the exception context to get back to the host.